File: prep/coreprep/graph/Kosaraju.java
import java.util.*;

public class Kosaraju {
  static class Graph {
    static class Vertex {
      Set<Integer> vs;

      Vertex() { this.vs = new HashSet<>(); }
      void addEdge(int v) { this.vs.add(v); }
      List<Integer> getAdjacentVertices() {
        List<Integer> neighbours = new ArrayList<>();

        for (int v : this.vs) {
          neighbours.add(v);
        }

        return neighbours;
      }
    }

    private int size;
    private List<Vertex> vertices;

    Graph(int size) {
      this.size = size;
      this.vertices = new ArrayList<>(size);

      for (int i = 0; i < size; i++) {
        this.vertices.add(new Vertex());
      }
    }

    void addEdge(int from, int to) {
      if (from < 0 || from >= this.size || to < 0 || to >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      this.vertices.get(from).addEdge(to);
    }

    List<Integer> getAdjacentVertices(int v) {
      if (v < 0 || v >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      return this.vertices.get(v).getAdjacentVertices();
    }

    int size() { return this.size; }

    Graph transpose() {
      Graph trans = new Graph(this.size);

      for (int i = 0; i < this.size; i++) {
        for (int v : getAdjacentVertices(i)) {
          trans.addEdge(v, i);
        }
      }

      return trans;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      Graph g = new Graph(n);

      int m = in.nextInt();
      for (int i = 0; i < m; i++) {
        int from = in.nextInt();
        int to = in.nextInt();
        g.addEdge(from, to);
      }

      kosaraju(g);
    }
  }

  // O(|V| + |E|)
  private static void kosaraju(Graph g) {
    Stack<Integer> st = new Stack<>();
    boolean[] visited = new boolean[g.size];

    for (int i = 0; i < g.size(); i++) {
      dfs1(g, visited, 0, st);
    }

    Graph trans = g.transpose();
    List<List<Integer>> scc = new ArrayList<>();

    visited = new boolean[g.size()];
    while (!st.empty()) {
      int v = st.pop();
      List<Integer> comp = new ArrayList<>();

      dfs2(trans, visited, v, comp);

      if (!comp.isEmpty()) {
        scc.add(comp);
      }
    }

    System.out.println(scc.size());
    if (!scc.isEmpty()) {
      for (List<Integer> comp : scc) {
        for (int v : comp) {
          System.out.printf("%d ", v);
        }
        System.out.println();
      }
    }
  }

  private static void dfs1(Graph g, boolean[] visited, int currVertex,
                           Stack<Integer> st) {
    if (visited[currVertex]) {
      return;
    }

    visited[currVertex] = true;

    for (int neighbour : g.getAdjacentVertices(currVertex)) {
      if (!visited[neighbour]) {
        dfs1(g, visited, neighbour, st);
      }
    }
    st.push(currVertex);
  }

  private static void dfs2(Graph g, boolean[] visited, int currVertex,
                           List<Integer> scc) {
    if (visited[currVertex]) {
      return;
    }

    visited[currVertex] = true;
    scc.add(currVertex);

    for (int neighbour : g.getAdjacentVertices(currVertex)) {
      if (!visited[neighbour]) {
        dfs2(g, visited, neighbour, scc);
      }
    }
  }
}

File: lc/coreprep/implementation/main_iterations/iteration_cpp/kosaraju.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

using namespace std;

struct Vertex {
  set<int> vs;

  void add_edge(int v) { this->vs.insert(v); }
  const vector<int> get_adjacent_vertices() const {
    vector<int> neighbours;

    for (int v : this->vs) {
      neighbours.push_back(v);
    }

    return neighbours;
  }
};

class Graph {
private:
  int sz;
  vector<Vertex> vertices;

public:
  Graph(int sz) : sz(sz) {
    for (int i = 0; i < sz; i++) {
      Vertex v;
      this->vertices.emplace_back(v);
    }
  }

  void add_edge(int from, int to) {
    if (from < 0 || from >= this->sz || to < 0 || to >= this->sz) {
      throw "illegal vertex";
    }

    this->vertices[from].add_edge(to);
  }

  const vector<int> get_adjacent_vertices(int v) const {
    if (v < 0 || v >= this->sz) {
      throw "illegal vertex";
    }

    return this->vertices[v].get_adjacent_vertices();
  }

  Graph transpose() {
    Graph trans(this->sz);

    for (int i = 0; i < this->sz; i++) {
      for (int v : get_adjacent_vertices(i)) {
        trans.add_edge(v, i);
      }
    }

    return trans;
  }

  int size() const { return this->sz; }
};

void dfs1(Graph &g, vector<bool> &visited, int curr_vertex, vector<int> &st) {
  if (visited[curr_vertex]) {
    return;
  }

  visited[curr_vertex] = true;

  for (int neighbour : g.get_adjacent_vertices(curr_vertex)) {
    if (!visited[neighbour]) {
      dfs1(g, visited, neighbour, st);
    }
  }

  st.push_back(curr_vertex);
}

void dfs2(Graph &g, vector<bool> &visited, int curr_vertex, vector<int> &comp) {
  if (visited[curr_vertex]) {
    return;
  }

  visited[curr_vertex] = true;
  comp.push_back(curr_vertex);

  for (int neighbour : g.get_adjacent_vertices(curr_vertex)) {
    if (!visited[neighbour]) {
      dfs2(g, visited, neighbour, comp);
    }
  }
}

void kosaraju(Graph &g) {
  vector<bool> visited(g.size());
  fill(visited.begin(), visited.end(), false);

  vector<int> st(g.size());

  for (int i = 0; i < g.size(); i++) {
    if (!visited[i]) {
      dfs1(g, visited, i, st);
    }
  }

  Graph trans = g.transpose();
  fill(visited.begin(), visited.end(), false);

  vector<vector<int>> scc;
  while (!st.empty()) {
    int v = st.back();
    st.pop_back();

    vector<int> comp;
    dfs2(trans, visited, v, comp);

    if (!comp.empty()) {
      scc.push_back(comp);
    }
  }

  cout << scc.size() << "\n";
  for (auto comp : scc) {
    for (auto c : comp) {
      cout << c << " ";
    }
    cout << "\n";
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, from, to;
  cin >> n >> m;

  Graph g(n);
  for (int i = 0; i < m; i++) {
    cin >> from >> to;
    g.add_edge(from, to);
  }

  kosaraju(g);

  return 0;
}